talisker/metrics.py

```diff
@@ -132,3 +139,175 @@ def inc(self, amount=1, **labels):
             client = talisker.statsd.get_client()
             name = self.get_statsd_name(labels)
             client.incr(name, amount)
+
+
+histogram_archive = 'histogram_archive.db'
+counter_archive = 'counter_archive.db'
+
+
+def histogram_sorter(sample):
+    # sort histogram samples in order of bucket size
+    name, labels, _ = sample
+    return name, float(labels.get('le', 0))
+
+
+def prometheus_cleanup_worker(pid):
+    """Clean up after a multiprocess worker has died."""
+    if prometheus_client is None:
+        return
+
+    try:
+        # just deletes delete gauge files
+        multiprocess.mark_process_dead(pid)
+        prom_dir = os.environ['prometheus_multiproc_dir']
+        worker_files = [
+            'histogram_{}.db'.format(pid),
+            'counter_{}.db'.format(pid),
+        ]
+        paths = [os.path.join(prom_dir, f) for f in worker_files]
+
+        # check at least one worker file exists
+        if not any(os.path.exists(path) for path in paths):
+            return
+
+        files = [
+            os.path.join(prom_dir, f) for f in
+            [histogram_archive, counter_archive] + worker_files
+        ]
+        metrics = collect(files)
+
+        tmp_histogram = tempfile.NamedTemporaryFile(delete=False)
+        tmp_counter = tempfile.NamedTemporaryFile(delete=False)
+        write_metrics(metrics, tmp_histogram.name, tmp_counter.name)
+
+        os.rename(
+            tmp_histogram.name, os.path.join(prom_dir, histogram_archive))
+        os.rename(
+            tmp_counter.name, os.path.join(prom_dir, counter_archive))
+
+        for path in paths:
+            if os.path.exists(path):
+                os.unlink(path)
+
+    except Exception:
+        # we should never fail at cleaning up
+        logger.exception('failed to cleanup prometheus worker files')
+
+
+def collect(files):
+    """This almost verbatim from MultiProcessCollector.collect().
+
+    It differs in a few ways:
+
+    1. it takes its files as an argument, rather than hardcoding '*.db', an
+       skips none existent files
+    2. it does not accumulate histograms, as it is intended for writing that
+       data back out
+    3. it uses an OrderedDict to preserve label order, and thus metric identity
+
+    It needs to be kept up to date with changes to prometheus_client as much as
+    possible, or until changes are landed upstream to allow better reuse.
+    """
```

Comment:
WDYT about an explicit whitelist and failure if we're using a version of prometheus_client that's not been checked?

It means we have to explicitly double check this when we bump prometheus_client but I think that's a Good Thing:tm:

Reply:
I think I would probably prefer to just issue a warning on startup. We can't control what versions of prometheus_client is used in the main project, as I wouldn't want to error or block if they it goes beyond what talisker supports.

Reply:
This code duplication looks dangerous to me; I'd rather fork `prometheus_client` instead. That said, can't why "unaccumulate" results as in previous versions of this branch?

Reply:
Once we've validated this approach in prod, I intend to propose some small changes upstream that will mean this duplication can disappear, time permitting.

Reply:
@maxiberta we can unaccumlate, and I did, but we cannot preserve label order without duplicating, and that is essential.

I don't think forking the whole library is a better solution that a custom collection function, that is only used for one specific purpose.

Reply:
If you look at the first commit, you can see an approach that did not duplicate. I got it 90% of the way there, only to fail at the final hurdle.

Reply:
ultimately, prometheus scraping is not sensitive to label order, but the mmaped file format parsing is